Fragments

mohamedfarith opened the pull request:
Have added exercise part in chapter fragments to this branch

mohamedfarith committed:
Added Fragment to activity

mohamedfarith committed:
Added shared prefences to login page

mohamedfarith committed:
Merge branch 'master' into Fragments

mohamedfarith committed:
Model login page name is changed to model login activity

nmrafiq87 commented on app/src/main/java/com/example/admin/learningandroid/fragments/SendingFragment.java:
```diff
@@ -0,0 +1,109 @@
+    public PassingData data;
+    public EditText etSendingFragment;
+    public Button buttonSendData;
+    public String editedText;
```
As discussed earlier if declaring a string use strText

nmrafiq87 commented on app/src/main/java/com/example/admin/learningandroid/fragments/SendingFragment.java:
```diff
@@ -0,0 +1,109 @@
+public class SendingFragment extends android.support.v4.app.Fragment {
+    private static final String mTAG = "Sending Fragment";
+    public PassingData data;
+    public EditText etSendingFragment;
```
More meaningful names etSendingFragment is confusing

nmrafiq87 commented on app/src/main/java/com/example/admin/learningandroid/fragments/SendingFragment.java:
```diff
@@ -0,0 +1,109 @@
+    }
+
+    @Override
+    public void onStart() {
```
Remove the lifecycle methods that are not required please

nmrafiq87 commented on app/src/main/java/com/example/admin/learningandroid/fragments/PassingData.java:
```diff
@@ -0,0 +1,7 @@
+
+import android.widget.EditText;
+
+interface PassingData {
```
public static class FragmentA extends ListFragment {
    ...
    // Container Activity must implement this interface
    public interface OnArticleSelectedListener {
        public void onArticleSelected(Uri articleUri);
    }
    ...
}

https://developer.android.com/guide/components/fragments

Please check the sample and follow naming conventions for an Interface accordingly.

nmrafiq87 commented on app/src/main/java/com/example/admin/learningandroid/fragments/PassingData.java:
```diff
@@ -0,0 +1,7 @@
+import android.widget.EditText;
+
+interface PassingData {
+    void changeData(String textData, EditText etSendingFragment);
```
Why do you need both textData and edit text at the same

nmrafiq87 commented on app/src/main/java/com/example/admin/learningandroid/fragments/ReceivingFragment.java:
```diff
@@ -0,0 +1,112 @@
+        // Required empty public constructor
+    }
+
+    public void getData(String data, final EditText sendingFragment) {
```
This edit text need not be passed from the interface. Use the activity to talk to other fragment to clear data

nmrafiq87 replied:
or it could be another interface as in other fragment

nmrafiq87 commented on app/src/main/java/com/example/admin/learningandroid/fragments/SendingFragment.java:
```diff
@@ -0,0 +1,109 @@
+        buttonSendData.setOnClickListener(new View.OnClickListener() {
+            @Override
+            public void onClick(View view) {
+                editedText = etSendingFragment.getText().toString();
```
As discussed pass only the string

nmrafiq87 commented on app/src/main/java/com/example/admin/learningandroid/layouts/ModelLoginActivity.java:
```diff
@@ -0,0 +1,108 @@
+import com.example.admin.learningandroid.MainActivity;
+import com.example.admin.learningandroid.R;
+
+public class ModelLoginPage extends AppCompatActivity {
```
The declaration says ModelLoginPage but the name is ModelLoginActivity. This will lead to compilation failure please run the code once before giving for review

nmrafiq87 commented on app/src/main/java/com/example/admin/learningandroid/layouts/ModelLoginActivity.java:
```diff
@@ -0,0 +1,108 @@
+            @Override
+            public void onClick(View view) {
+                //To make only one time sign up
+                SharedPreferences firstRetrieve = getSharedPreferences("SAVE_DATA", MODE_PRIVATE);
```
Meaningful name pls for instance userPreference

nmrafiq87 replied:
Please use constants for SAVE_DATA as already discussed in the previous reviews

nmrafiq87 commented on app/src/main/java/com/example/admin/learningandroid/layouts/ModelLoginActivity.java:
```diff
@@ -0,0 +1,108 @@
+            public void onClick(View view) {
+                //To make only one time sign up
+                SharedPreferences firstRetrieve = getSharedPreferences("SAVE_DATA", MODE_PRIVATE);
+                String savedName = firstRetrieve.getString("NAME", "");
```
Again use constants for NAME

nmrafiq87 commented on app/src/main/java/com/example/admin/learningandroid/layouts/ModelLoginActivity.java:
```diff
@@ -0,0 +1,108 @@
+                //To make only one time sign up
+                SharedPreferences firstRetrieve = getSharedPreferences("SAVE_DATA", MODE_PRIVATE);
+                String savedName = firstRetrieve.getString("NAME", "");
+                String savedRegNumber = firstRetrieve.getString("REGISTRATION_NUMBER", "");
```
Same here

nmrafiq87 commented on app/src/main/java/com/example/admin/learningandroid/layouts/ModelLoginActivity.java:
```diff
@@ -0,0 +1,108 @@
+    public Button btnLogin;
+    String mName;
+    String mRegistrationNumber;
+    public String name;
```
Why are there too many names ?

nmrafiq87 commented on app/src/main/java/com/example/admin/learningandroid/layouts/ModelLoginActivity.java:
```diff
@@ -0,0 +1,108 @@
+
+                int mCount = mName.length();
+                //condition for various credential check
+                if (mCount <= 3) {
```
The following statement if - else block is very complicated
                String message = "";
                etName.setText(firstRetrieve.getString("NAME", ""));
                etRegistrationNumber.setText(firstRetrieve.getString("REGISTRATION_NUMBER", ""));
                mName = etName.getText().toString().trim();
                mRegistrationNumber = etRegistrationNumber.getText().toString().trim();
                if (mName.length() <= 3){
                    message = "Enter valid name";
                }else if (TextUtils.isEmpty(mRegistrationNumber)){
                    message = "Enter valid register number";
                }else{
                    message = "Saved successfully";
                }

                Toast.makeText(ModelLoginActivity.this, message, Toast.LENGTH_SHORT).show();

nmrafiq87 commented on app/src/main/java/com/example/admin/learningandroid/layouts/ModelLoginActivity.java:
```diff
@@ -0,0 +1,108 @@
+                String savedName = firstRetrieve.getString("NAME", "");
+                String savedRegNumber = firstRetrieve.getString("REGISTRATION_NUMBER", "");
+
+                mName = etName.getText().toString();
```
use etName.getText().toString().trim()

nmrafiq87 commented on app/src/main/java/com/example/admin/learningandroid/layouts/ModelLoginActivity.java:
```diff
@@ -0,0 +1,108 @@
+                String registrationNumber = preference.getString("REGISTRATION_NUMBER", "");
+
+             //making sure that the user do not logs in with empty data
+                if (name!="" && registrationNumber!="") {
```
Try simplifying this code as well can be better structured for simplicity and readability. If Else is very convoluted

nmrafiq87 commented on app/src/main/res/layout/sending_fragment_layout.xml:
```diff
@@ -0,0 +1,24 @@
+        android:layout_height="wrap_content"
+        android:layout_margin="10dp"
+        android:layout_gravity="center"
+        android:text="click to send data"/>
```
Try to use resources for strings

nmrafiq87 commented on app/src/main/res/layout/receiving_fragment.xml:
```diff
@@ -0,0 +1,26 @@
+        android:layout_width="200dp"
+        android:layout_height="60dp"
+        android:layout_margin="20dp"
+        android:text="clear text"
```
use string resources

nmrafiq87 commented on app/src/main/res/layout/receiving_fragment.xml:
```diff
@@ -0,0 +1,26 @@
+    android:layout_width="match_parent"
+    android:layout_height="match_parent"
+    android:orientation="vertical"
+    android:background="#CE93D8"
```
use colors

nmrafiq87 commented on app/src/main/java/com/example/admin/learningandroid/layouts/ModelLoginActivity.java:
```diff
@@ -0,0 +1,108 @@
+                        Toast.makeText(ModelLoginPage.this, "enter valid register number", Toast.LENGTH_SHORT).show();
+                    } else {
+                        Toast.makeText(ModelLoginPage.this, "Successfully saved", Toast.LENGTH_SHORT).show();
+                        if (savedName.equals("") && savedRegNumber.equals("")) {
```
Use TextUtils.isEmpty which does the same

nmrafiq87 reviewed (changes requested):
Please make the required changes

mohamedfarith committed:
made changes as per suggestions

mohamedfarith committed:
-->Resolved conflicts while updating master with branch Fragments
-->Merge branch 'master' into Fragments

nmrafiq87 commented on app/src/main/AndroidManifest.xml:
```diff
@@ -63,7 +63,14 @@
         <activity android:name=".layouts.RelativeLayoutActivity" />
         <activity android:name=".layouts.ConstraintLayoutActivity" />
         <activity android:name=".layouts.LinearLayoutActivity" />
+<<<<<<< HEAD
```
Back merge not done properly

nmrafiq87 commented on app/src/main/java/com/example/admin/learningandroid/fragments/FragmentActivity.java:
```diff
@@ -0,0 +1,41 @@
+import com.example.admin.learningandroid.R;
+
+
+public class FragmentActivity extends AppCompatActivity implements PassingDataFromSendingFragmentToReceivingFragment,PassingDataFromReceivingFragmentToSendingFragment {
```
I think the names should be shorter and crisper and should follow the conventions

nmrafiq87 replied:
The interface can have two methods. The name of the interface can be something like FragmentInterface and it can have two methods like setData and clearData

nmrafiq87 commented on app/src/main/java/com/example/admin/learningandroid/fragments/FragmentActivity.java:
```diff
@@ -0,0 +1,41 @@
+        super.onCreate(savedInstanceState);
+        setContentView(R.layout.fragment_layout);
+        Log.v(TAG, "onCreate of Fragment Activity");
+
```
Avoid line breaks

nmrafiq87 commented on app/src/main/java/com/example/admin/learningandroid/fragments/PassingDataFromReceivingFragmentToSendingFragment.java:
```diff
@@ -0,0 +1,7 @@
+
+public interface PassingDataFromReceivingFragmentToSendingFragment {
+
+    void passingDataFromReceivingFragmentToSendingFragment();
```
As i said there can be one interface and two methods

mohamedfarith committed:
-->Interface and its method names changed.
-->Deleted empty spaces.
-->checked for project build.

mohamedfarith committed:
*Issues with the previous commit:
 -->While back merging master with fragments conflicts arouse,while resolving conflict ModelLoginActivity as in master replaced
 that in Fragments which had shared preference added.
 *Reslved Issues:
 -->Added shared preference to the ModelLoginActivity.
 -->removed empty spaces.

nmrafiq87 commented on app/src/main/java/com/example/admin/learningandroid/layouts/ModelLoginActivity.java:
```diff
@@ -9,40 +10,78 @@
+                Toast.makeText(ModelLoginActivity.this, message, Toast.LENGTH_SHORT).show();
+                break;
+            case R.id.btn_login:
+                SharedPreferences preference = getSharedPreferences(SAVE_DATA, MODE_PRIVATE);
```
Lot of duplicate declarations for preference. Can you please initialise once at the start of the method. The same for the name and registration number. You could even initialize during on create of the activity

nmrafiq87 commented on app/src/main/java/com/example/admin/learningandroid/layouts/ModelLoginActivity.java:
```diff
@@ -9,40 +10,78 @@
+                String registrationNumber = preference.getString(REGISTRATION_NUMBER, "");
+                //making sure that the user do not logs in with empty data
+
+                if (!name.isEmpty() && !registrationNumber.isEmpty()) {
```
Use TextUtils.isEmpty rather than isEmpty and figure out why i am suggesting the same

mohamedfarith committed:
-->As all the files under ".idea/" should be under gitignore file,added .idea/caches/build_file_checksums
-->Duplicate variables for SharedPreference,name and registration number is removed.
-->isEmpty is replaced by TextUtils.isEmpty.
